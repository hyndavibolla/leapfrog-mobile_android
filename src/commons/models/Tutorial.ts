import { ReactNode } from 'react';
import { StyleProp, ViewStyle } from 'react-native';

export interface Tutorial {
  isParent?: boolean;
  isHidden?: boolean;
  parentId?: number;
  scrollToDown?: number;
}

export interface TutorialElementPosition extends Tutorial {
  x: number;
  y: number;
  children: ReactNode;
  style: StyleProp<ViewStyle>;
}
