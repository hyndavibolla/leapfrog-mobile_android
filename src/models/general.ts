import { Dispatch } from 'react';
import Auth0 from 'react-native-auth0';

import { ApiService } from '_services/ApiService';
import { AppService } from '_services/AppService';
import { CCPAService } from '_services/CCPAService';
import { ICrashlyticsService } from '_services/CrashlyticsService';
import { EventTrackerService } from '_services/EventTrackerService';
import { IHttpService } from '_services/HttpService';
import { NativeHelperService } from '_services/NativeHelperService';
import { Logger } from '_services/Logger';
import { IPerformanceTrackingService } from '_services/PerformanceTrackingService';
import { RaiseService } from '_services/RaiseService';
import { RemoteConfigService } from '_services/RemoteConfigService';
import { ICardLinkState } from '_state_mgmt/cardLink/state';
import { ICoreState } from '_state_mgmt/core/state';
import { IGameState } from '_state_mgmt/game/state';
import { IGiftCardState } from '_state_mgmt/giftCard/state';
import { IMessagesState } from '_state_mgmt/messages/state';
import { IMissionState } from '_state_mgmt/mission/state';
import { IRewardState } from '_state_mgmt/reward/state';
import { ISurveyState } from '_state_mgmt/survey/state';
import { IUserState } from '_state_mgmt/user/state';

export interface IAction<T = string, P = any> {
  type: T;
  payload: P;
}

export type IReducer<S = any, A = any> = (state: S, action: IAction<A>) => S;

export interface Deps {
  apiService: ApiService;
  appService: AppService;
  crashlyticsService: ICrashlyticsService;
  httpService: IHttpService;
  performanceTrackingService: IPerformanceTrackingService;
  raiseService: RaiseService;
  nativeHelperService: NativeHelperService;
  stateSnapshot: { get: () => IGlobalState; set: (state: IGlobalState) => void };
  logger: Logger;
  eventTrackerService: EventTrackerService;
  ccpaService: CCPAService;
  remoteConfigService: RemoteConfigService;
  authService: Auth0;
}
export type IGlobalState = {
  core: ICoreState;
  user: IUserState;
  game: IGameState;
  mission: IMissionState;
  reward: IRewardState;
  survey: ISurveyState;
  messages: IMessagesState;
  cardLink: ICardLinkState;
  giftCard: IGiftCardState;
};

export interface IGlobalContext {
  state: IGlobalState;
  dispatch: Dispatch<IAction<any, any>>;
  deps: Deps;
}

export enum HTTP_METHOD {
  GET = 'GET',
  POST = 'POST',
  PUT = 'PUT',
  PATCH = 'PATCH',
  DELETE = 'DELETE'
}

export interface ITokenSet {
  accessToken: string;
  refreshToken: string;
  accessTokenExpiryTime: string;
  sywToken?: string;
  /**
   * @deprecated deviceId does not belong to ITokenSet anymore.
   */
  deviceId?: string;
  scope?: string;
  expiresIn?: number;
  tokenType?: string;
}

export enum FeatureFlag {
  API_OVERRIDE = 'API_OVERRIDE',
  TEST_ONLY = 'TEST_ONLY',
  ENV_BANNER = 'ENV_BANNER',
  BUTTON_DEBUG = 'BUTTON_DEBUG',
  LOG_LEVEL_INFO = 'LOG_LEVEL_INFO',
  LOG_LEVEL_DEBUG = 'LOG_LEVEL_DEBUG',
  LOG_LEVEL_WARN = 'LOG_LEVEL_WARN',
  LOG_LEVEL_ERROR = 'LOG_LEVEL_ERROR',
  LOG_VIEW = 'LOG_VIEW',
  STREAK_REDEEM_CARD = 'STREAK_REDEEM_CARD',
  SAILTHRU_TRACKING = 'SAILTHRU_TRACKING',
  SURVEY = 'SURVEY',
  PREDICTIVE_MAP_SEARCH = 'PREDICTIVE_MAP_SEARCH',
  MISSIONS_SEARCH = 'MISSIONS_SEARCH',
  SEE_ALL_TRANSACTIONS = 'SEE_ALL_TRANSACTIONS',
  WALLET_YOUR_GC = 'WALLET_YOUR_GC'
}

export type DateLike = Date | string | number;

export type HTMLString = string;

export interface IGameLevelUI {
  updatedAt: DateLike;
  level: number;
}

export interface ICCPASetting {
  allow: boolean;
}

export interface IOnboarding {
  seenOnboarding: boolean;
}

export enum ButtonCreativeType {
  HERO = 'hero',
  CAROUSEL = 'carousel',
  LIST = 'list',
  GRID = 'grid',
  DETAIL = 'detail',
  OTHER = 'other'
}

export interface ISearchHistoryItem {
  keyword: string;
  lastUpdatedAt: DateLike;
}

export enum TooltipKey {
  EARN = 'earn',
  REWARDS = 'rewards',
  MISSIONS = 'missions',
  STREAK = 'streak'
}

export interface IApiOverrideRule {
  key: string;
  description?: string;
  matcher: string; // reg exp
  status: number;
  response: any;
}

export namespace RemoteConfigModel {
  export interface IAppVersion {
    suggested: number;
    required: number;
  }

  export interface IEarnSection {
    name: string;
    order: number;
    visible: boolean;
    title: string;
    description: string;
    seeAllButton: boolean;
  }
}

export enum RUNTIME_ENV {
  LIVE = 'live',
  DEV_TEST = 'test'
}

export interface IMeasure {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface IExpirePointsBannerDataSet {
  lastPointsDate: DateLike;
  lastAcceptedDate: DateLike;
}

export interface ISailthruMessage {
  createdAt: string;
  custom: {
    [k: string]: string;
  };
  htmlText: string;
  id: string;
  isRead: boolean;
  text: string;
  title: string;
  type: string;
  cardImageUrl?: string;
  share: boolean;
}

export interface ITutorialStatus {
  isBannerWatched: boolean;
  date: string;
}
