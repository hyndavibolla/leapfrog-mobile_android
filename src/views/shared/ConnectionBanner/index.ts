export { default as ConnectionBanner } from './ConnectionBanner';
