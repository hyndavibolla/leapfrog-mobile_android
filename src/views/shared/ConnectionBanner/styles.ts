import { css } from '@emotion/native';
import { COLOR, FONT_SIZE } from '../../../constants';

export const styles = {
  container: css`
    flex-direction: row;
    background-color: ${COLOR.ORANGE};
    padding: 8px 16px;
    align-items: center;
  `,
  title: css`
    color: ${COLOR.WHITE};
    font-size: ${FONT_SIZE.SMALL};
  `,
  icon: css`
    margin-right: 8px;
  `
};
