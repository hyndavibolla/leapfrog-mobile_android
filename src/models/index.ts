import * as GeneralModel from './general';
import * as UserModel from './user';
import * as ActivityModel from './activity';
import * as OfferModel from './offer';
import * as GameModel from './game';
import * as MissionModel from './mission';
import * as RewardModel from './reward';
import * as StreakModel from './streak';
import * as PromptModel from './prompt';
import * as CardLink from './cardLink';

export { GeneralModel, UserModel, ActivityModel, OfferModel, GameModel, MissionModel, RewardModel, StreakModel, PromptModel, CardLink };
