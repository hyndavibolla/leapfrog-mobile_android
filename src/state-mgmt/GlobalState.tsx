import React, { createContext, memo, PropsWithChildren, useReducer, Dispatch, useEffect } from 'react';

import { IGlobalState, Deps, IReducer, IGlobalContext } from '_models/general';
import { reducer as coreReducer } from './core/reducer';
import { reducer as userReducer } from './user/reducer';
import { reducer as gameReducer } from './game/reducer';
import { reducer as missionReducer } from './mission/reducer';
import { reducer as rewardReducer } from './reward/reducer';
import { reducer as surveyReducer } from './survey/reducer';
import { reducer as messagesReducer } from './messages/reducer';
import { reducer as cardLinkReducer } from './cardLink/reducer';
import { reducer as giftCardReducer } from './giftCard/reducer';
import { initialState as coreInitialState } from './core/state';
import { initialState as userInitialState } from './user/state';
import { initialState as gameInitialState } from './game/state';
import { initialState as missionInitialState } from './mission/state';
import { initialState as rewardInitialState } from './reward/state';
import { initialState as surveyInitialState } from './survey/state';
import { initialState as messagesInitialState } from './messages/state';
import { initialState as cardLinkInitialState } from './cardLink/state';
import { initialState as giftCardInitialState } from './giftCard/state';
import { combineReducers } from '_utils/combineReducers';
import { GeneralModel } from '_models';
import { deepClone } from '_utils/deepClone';

export const getInitialState = () =>
  deepClone<IGlobalState>({
    core: coreInitialState,
    user: userInitialState,
    game: gameInitialState,
    mission: missionInitialState,
    reward: rewardInitialState,
    survey: surveyInitialState,
    messages: messagesInitialState,
    cardLink: cardLinkInitialState,
    giftCard: giftCardInitialState
  });

export const combinedReducer = combineReducers<GeneralModel.IGlobalState>({
  core: coreReducer,
  user: userReducer,
  game: gameReducer,
  mission: missionReducer,
  reward: rewardReducer,
  survey: surveyReducer,
  messages: messagesReducer,
  cardLink: cardLinkReducer,
  giftCard: giftCardReducer
});

export const GlobalContext = createContext<{ state: IGlobalState; dispatch: Dispatch<any>; deps: Deps }>({
  state: getInitialState(),
  dispatch: /* istanbul ignore next */ () => null,
  deps: null
} as IGlobalContext);

export const GlobalProvider = memo(
  ({
    children,
    deps,
    initState,
    combinedReducers
  }: PropsWithChildren<{
    deps: Deps;
    initState: IGlobalState;
    combinedReducers: IReducer<IGlobalState>;
  }>) => {
    const [state, dispatch] = useReducer(combinedReducers, initState);
    useEffect(() => deps.stateSnapshot.set(state), [state]); // eslint-disable-line react-hooks/exhaustive-deps
    return <GlobalContext.Provider value={{ state, dispatch, deps } as IGlobalContext}>{children}</GlobalContext.Provider>;
  }
);
