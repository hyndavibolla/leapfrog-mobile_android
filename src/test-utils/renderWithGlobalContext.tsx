import React, { ReactElement } from 'react';
import { render } from '@testing-library/react-native';

import { getInitialState, GlobalProvider } from '../state-mgmt/GlobalState';
import { Deps, IGlobalState, IReducer } from '../models/general';
import { getMockDeps } from './getMockDeps';
import { getMockReducer } from './getMockReducer';

export const renderWithGlobalContext = (
  children: ReactElement,
  deps: Deps = getMockDeps(),
  initState: IGlobalState = getInitialState(),
  combinedReducers: IReducer<IGlobalState, any> = getMockReducer()
) => {
  const renderAPI = render(
    <GlobalProvider deps={deps} initState={initState} combinedReducers={combinedReducers}>
      {children}
    </GlobalProvider>
  );
  return { ...renderAPI, mockReducer: combinedReducers, deps };
};
