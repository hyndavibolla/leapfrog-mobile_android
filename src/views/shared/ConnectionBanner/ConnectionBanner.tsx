import React, { memo, useContext } from 'react';
import { StyleProp, View, ViewStyle } from 'react-native';

import { useTestingHelper } from '_utils/useTestingHelper';
import { GlobalContext } from '_state_mgmt/GlobalState';
import { COLOR, FONT_FAMILY, ICON } from '_constants';
import { styles } from './styles';
import { Icon } from '_commons/components/atoms/Icon';

import { Text } from '../Text';

export interface Props {
  style?: StyleProp<ViewStyle>;
  title?: string;
}

export const ConnectionBanner = ({ style, title = "You don't have an internet connection" }: Props) => {
  const { getTestIdProps } = useTestingHelper('connection-banner');

  const { state } = useContext(GlobalContext);
  return state.core.isConnected ? null : (
    <View style={[styles.container, style]} {...getTestIdProps('container')}>
      <Icon name={ICON.NO_CONNECTION} color={COLOR.WHITE} style={styles.icon} />
      <Text font={FONT_FAMILY.BOLD} style={styles.title}>
        {title}
      </Text>
    </View>
  );
};

export default memo(ConnectionBanner);
