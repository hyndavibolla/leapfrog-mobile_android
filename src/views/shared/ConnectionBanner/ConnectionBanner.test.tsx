import React from 'react';

import ConnectionBanner, { Props } from './ConnectionBanner';
import { renderWithGlobalContext } from '../../../test-utils/renderWithGlobalContext';
import { getInitialState } from '../../../state-mgmt/GlobalState';
import { IGlobalState } from '../../../models/general';

describe('ConnectionBanner', () => {
  let props: Props;
  let initialState: IGlobalState;

  beforeEach(() => {
    initialState = getInitialState();
    props = {
      title: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry'
    };
  });

  it('should render with connection', () => {
    const { toJSON } = renderWithGlobalContext(<ConnectionBanner {...props} />, undefined, {
      ...initialState,
      core: { ...initialState.core, isConnected: true }
    });
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render WITHOUT connection', () => {
    const { toJSON } = renderWithGlobalContext(<ConnectionBanner {...props} />, undefined, {
      ...initialState,
      core: { ...initialState.core, isConnected: false }
    });
    expect(toJSON()).toMatchSnapshot();
  });
});
