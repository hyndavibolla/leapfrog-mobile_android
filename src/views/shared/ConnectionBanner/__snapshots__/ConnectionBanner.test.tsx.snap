// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectionBanner should render WITHOUT connection 1`] = `
<View
  accessibilityLabel="connection-banner-container"
  accessible={true}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#FF5F00",
        "flexDirection": "row",
        "paddingBottom": 8,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 8,
      },
      undefined,
    ]
  }
  testID="connection-banner-container"
>
  <Text
    accessibilityLabel="icon-text"
    accessible={true}
    allowFontScaling={false}
    selectable={false}
    style={
      Array [
        Object {
          "color": "#FFFFFF",
          "fontSize": 20,
        },
        Object {
          "marginRight": 8,
        },
        Object {
          "fontFamily": "sywmax",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
    testID="icon-text"
  >
    î¤±
  </Text>
  <Text
    font="SFProDisplay-Bold"
    style={
      Array [
        Object {
          "color": "#909BA0",
          "fontFamily": "SFProDisplay-Bold",
          "fontSize": 16,
          "lineHeight": 22,
        },
        Object {
          "color": "#FFFFFF",
          "fontSize": 18,
        },
      ]
    }
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry
  </Text>
</View>
`;

exports[`ConnectionBanner should render with connection 1`] = `null`;
